drivers/interrupt_controller/intc_ioapic.c

```diff
@@ -125,28 +125,65 @@ static const struct device *const vtd =
 	DEVICE_DT_GET_OR_NULL(DT_INST(0, intel_vt_d));
 static uint16_t ioapic_id;
 
-static bool get_vtd(void)
+static void devscope_handler(ACPI_DMAR_DEVICE_SCOPE *devscope, void *arg)
 {
-	union acpi_dmar_id dmar_id;
-	int inst_cnt;
+	ACPI_DMAR_PCI_PATH *dev_path;
+	union acpi_dmar_id pci_path;
 
-	if (!device_is_ready(vtd)) {
-		return false;
+	ARG_UNUSED(arg);
+
+	if (devscope->EntryType != ACPI_DMAR_SCOPE_TYPE_IOAPIC) {
+		return;
 	}
 
-	if (ioapic_id != 0) {
-		return true;
+	dev_path = ACPI_ADD_PTR(ACPI_DMAR_PCI_PATH, devscope,
+				sizeof(ACPI_DMAR_DEVICE_SCOPE));
+
+	/* Get first entry */
+	pci_path.bits.bus = devscope->Bus;
+	pci_path.bits.device = dev_path->Device;
+	pci_path.bits.function = dev_path->Function;
+
+	ioapic_id = pci_path.raw;
+}
+
+static void subtable_handler(ACPI_DMAR_HEADER *subtable, void *arg)
+{
+	ARG_UNUSED(arg);
+
+	if (subtable->Type == ACPI_DMAR_TYPE_HARDWARE_UNIT) {
+		ACPI_DMAR_HARDWARE_UNIT *hu;
+
+		hu = CONTAINER_OF(subtable, ACPI_DMAR_HARDWARE_UNIT, Header);
+		acpi_dmar_foreach_devscope(hu, devscope_handler, arg);
 	}
+}
 
-	/* Assume only one PCH in system (say client platform). */
-	if (!acpi_drhd_get(ACPI_DMAR_SCOPE_TYPE_IOAPIC, NULL, &dmar_id, &inst_cnt, 1u)) {
+static bool get_ioapic(void)
+{
+	ACPI_TABLE_DMAR *dmar = acpi_table_get("DMAR", 0);
+
```

Comment:
I don't like seeing all this ACPI related code into ioapic.

It should be minimal as possible. So move as much as possible into lib/acpi/

Reply:
moved to acpi lib